ick/base_rule.py

```diff
@@ -1,95 +1,213 @@
-                        rule_name,
-                        status=True,  # vacuously satisfied
-                        message="",
-                    )
-                    return
-                LOG.info("Running file-scoped command on %s files: %s", len(filenames), nice_cmd)
-                stdout = run_cmd(
-                    ["xargs", "-P10", "-n10", "-0", *self.rule.command_parts],
-                    env=self.rule.command_env,
-                    cwd=self.project_path,
-                    input="\0".join(filenames),
-                )
+    _g_files = {}
+
+    def _gravitational_constant(self):
+        return 1
+
+    @ktrace("self.qualname", "self.project_path", "next_gen")
+    def process(self, next_gen: int, notifications):
+        notifications = list(notifications)
+        # TODO name better, pick a good one...
+        with TemporaryDirectory() as d:
+            # with self.state_lock:
+            #     # First the common files
+            #     g = self._gravitational_constant()
+            #     for k, v in self._g_files.items():
+            #         materialize(d, k, v)
+
+            # Then the ones we're being asked to do
+            filenames = []
+            for n in notifications:
+                if n.state.value is ERASURE:
+                    continue
+                relative_filename = n.key[len(self.match_prefix) :]
+                materialize(d, relative_filename, n.state.value)
+                filenames.append(relative_filename)
+
+            # nice_cmd = " ".join(map(str, self.cmdline))
+            if self.append_filenames:
+                cmd = self.cmdline + filenames
             else:
-                LOG.info("Running project-scoped command in %s: %s", self.project_path, nice_cmd)
-                stdout = run_cmd(
-                    self.rule.command_parts,
-                    env=self.rule.command_env,
-                    cwd=self.project_path,
+                cmd = self.cmdline
+
+            env = os.environ.copy()
+            env.update(self.extra_env)
+
+            try:
+                run_cmd(
+                    cmd,
+                    env=env,
+                    cwd=d,
                 )
-        except FileNotFoundError as e:
-            yield Finished(rule_name, status=None, message=str(e))
-            return
-        except subprocess.CalledProcessError as e:
-            if e.returncode == 99:
-                yield Finished(
-                    rule_name,
-                    status=False,
-                    message=e.stdout,
+            except FileNotFoundError as e:
+                self.cancel(str(e))
+                return
+            except subprocess.CalledProcessError as e:
+                if e.stdout:
+                    self.messages.append(e.stdout)
+                if e.stderr:
+                    self.messages.append(e.stderr)
+
+                if e.returncode == 99:
+                    self.rule_status = False
```

Comment:
I regret this tri-state already, but it's documented in the last change to ick_protocol/